File: 1984-minimum-difference-between-highest-and-lowest-of-k-scores.py
class Solution:
    def minimumDifference(self, nums: list[int], k: int) -> int:
        '''
        We first sort the nums so each group of k-elements are close to each other.
        This ensures that the difference between highest and lowest is closest if we choose the contiguous groups.
        Next, we just use sliding window and calculate the differences and get the min
        '''
        nums.sort()
        min_diff = float('inf')
        for i in range(len(nums) - k + 1):
            curr_diff = nums[i+k-1] - nums[i]
            if curr_diff < min_diff:
                min_diff = curr_diff

        return min_diff


sol = Solution()
print(f'output: {sol.minimumDifference(nums=[90], k=1)}, expected: 0')
print(f'output: {sol.minimumDifference(nums=[9, 4, 1, 7], k=2)}, expected: 2')

File: 1984-minimum-difference-between-highest-and-lowest-of-k-scores.java
import java.util.Arrays;

class MinimumDifferenceBetweenHighestAndLowestOfKScores {
    public int minimumDifference(int[] nums, int k) {
        int minDiff = Integer.MAX_VALUE;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - k + 1; i++) {
            int currDiff = nums[i+k-1] - nums[i];
            if (currDiff < minDiff) {
                minDiff = currDiff;
            }
        }
        return minDiff;
    }

    public static void main(String[] args) {
        MinimumDifferenceBetweenHighestAndLowestOfKScores sol = new MinimumDifferenceBetweenHighestAndLowestOfKScores();
        System.out.println("Output: " + sol.minimumDifference(new int[] { 90 }, 1) + ", expected: 0");
        System.out.println("Output: " + sol.minimumDifference(new int[] { 9, 4, 1, 7 }, 2) + ", expected: 2");
    }
}